src/main/cpp/libclp_ffi_java/Java_MessageEncoder.cpp

```diff
@@ -146,68 +120,84 @@ JNIEXPORT void JNICALL Java_com_yscope_clp_compressorfrontend_MessageEncoder_enc
+        jni_env->DeleteLocalRef(local_class_ref);
+        throw JavaRuntimeException(__FILENAME__, __LINE__, jni_env, "NewGlobalRef failed");
+    }
+    jni_env->DeleteLocalRef(local_class_ref);
+
+    return global_class_ref;
+}
+
+static void encode_message_native (JNIEnv* jni_env, jbyteArray Java_message,
+                                   jobject Java_encodedMessage)
+{
     // Get the message
     auto message_bytes = get_java_primitive_array_elements<jbyteArray, jbyte>(
             jni_env, Java_message, JNI_ABORT);
-    if (nullptr == message_bytes) {
-        return;
-    }
     auto message_length = jni_env->GetArrayLength(Java_message);
     string_view message(size_checked_pointer_cast<char>(message_bytes.get()), message_length);
 
-    std::string logtype;
-    std::vector<encoded_variable_t> encoded_vars;
-    std::vector<int32_t> dictionary_var_bounds;
+    string logtype;
+    vector<encoded_variable_t> encoded_vars;
+    vector<int32_t> dictionary_var_bounds;
     if (false == encode_message(message, logtype, encoded_vars, dictionary_var_bounds)) {
-        JavaIOException::throw_in_java(jni_env, "message contains variable placeholders");
-        return;
+        throw JavaIOException(__FILENAME__, __LINE__, jni_env,
+                              "message contains variable placeholders");
     }
 
     // Set encodedMessage.logtype
-    if (logtype.length() > cJSizeMax) {
-        JavaIOException::throw_in_java(jni_env, "logtype too long for byte array");
-        return;
-    }
-    auto Java_logtype = jni_env->NewByteArray(static_cast<jsize>(logtype.length()));
-    if (nullptr == Java_logtype) {
-        return;
-    }
-    jni_env->SetByteArrayRegion(Java_logtype, 0, static_cast<jsize>(logtype.length()),
-                                size_checked_pointer_cast<jbyte>(logtype.data()));
-    auto exception = jni_env->ExceptionOccurred();
-    if (nullptr != exception) {
-        return;
-    }
+    auto Java_logtype = new_java_primitive_array<jbyteArray, jbyte>(
+            jni_env, size_checked_pointer_cast<jbyte>(logtype.data()), logtype.length());
     jni_env->SetObjectField(Java_encodedMessage, Java_EncodedMessage_logtype, Java_logtype);
 
     // Set encodedMessage.dictionaryVarBounds
     if (false == dictionary_var_bounds.empty()) {
-        if (dictionary_var_bounds.size() > cJSizeMax) {
-            JavaIOException::throw_in_java(jni_env, "dictionaryVarBounds can't fit in Java int "
-                                                    "array");
-            return;
-        }
-        auto java_dict_var_bounds =
-                jni_env->NewIntArray(static_cast<jsize>(dictionary_var_bounds.size()));
-        jni_env->SetIntArrayRegion(java_dict_var_bounds, 0,
-                                   static_cast<jsize>(dictionary_var_bounds.size()),
-                                   dictionary_var_bounds.data());
+        auto Java_dictVarBounds = new_java_primitive_array<jintArray, jint>(
```

Comment:
why Pascal case? Looks like there is an assumption for java related variables?

Reply:
See `README.md`